### PR TITLE
[pytorch] Fix pushLong() issue in pickler.

### PR DESCRIPTION
**Pull request by jjlilley**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28057 [pytorch] Fix pushLong() issue in pickler.**

For pushLong() in Pickler, it looks like we only use for a single use case, with a 10-byte value.

We were handling > 256 bytes incorrectly, by using a LONG4 opcode (expecting 4-byte length), but pushing 8 bytes. We could harden this handling, but rather than improve codepaths that we never expect to use, this change simply removes the incorrect codepath and adds and assert.

Differential Revision: [D17934174](https://our.internmc.facebook.com/intern/diff/D17934174/)

### CODE REVIEW

**Review by zdevito — CHANGES_REQUESTED** [review]
LONG4 is suppose to first read 4 bytes to determine the size of the data to read and then read that many bytes. I am not sure what our implementation does (either before or after the patch). LONG1 looks more correct, though even that only supports numbers that are precisely 8 bytes. @driazati for what is suppose to happen here. I think we should just delete LONG4 handling, and assert in LONG1 handling that length == 8 on write (currently we only do it on read).

**Comment by jjlilley** [comment]
Thanks for the feedback.                                                                                                                                         
I'm obviously not the expert in pickle opcodes, mostly noticed this in the other change and wanted to avoid an incorrect-size read.                              
AFAICT, it looks like there's only one caller of the pushLong function, and it's called with a 10-byte magic number, so asserting size==8 on writing a LONG1 isn\
't going to universally work.

**Comment by driazati** [comment]
`LONG1` of size 10 is needed for #25109, but I don't think we have a real use case for `LONG4` in the pickler, since TorchScript's ints are all `int64_t`s. I think we should just remove the `LONG4` handling and error out if the code tries to go down that path.

**Comment by jjlilley** [comment]
In any case, closing this specific PR.

**Comment by jjlilley** [comment]
> `LONG1` of size 10 is needed for #25109, but I don't think we have a real use case for `LONG4` in the pickler, since TorchScript's ints are all `int64_t`s. I think we should just remove the `LONG4` handling and error out if the code tries to go down that path.

That sounds reasonable, uploading a version that simply removes the incorrect codepath and asserts that the size fits in 1 byte.

**Review by zdevito — APPROVED** [review]

File: torch/csrc/jit/pickler.cpp [hunk 1]
```diff
@@ -415,17 +415,11 @@ void Pickler::pushDouble(double value) {
...
-    TORCH_INTERNAL_ASSERT(
-        data.size() > std::numeric_limits<uint32_t>::max(),
-        "Cannot pickle a long with a size larger than 4 bytes")
-    push<PickleOpCode>(PickleOpCode::LONG4);
-    // TODO: should this be uint32_t instead?
-    push<uint64_t>(size);
-  }
+  TORCH_INTERNAL_ASSERT(
+    size <= std::numeric_limits<uint8_t>::max(),
+    "Cannot pickle a long with a size larger than 256 bytes");
```
Comment (driazati) [inline torch/csrc/jit/pickler.cpp]: This should say `Cannot pickle a long with a size larger than 1 byte`

Comment (jjlilley) [reply torch/csrc/jit/pickler.cpp]: I can change this message, but given that data.size() can be up to 256, saying that the long's size can't be larger than 1 byte doesn't seem accurate.

I could say the more-awkward-but-accurate:
    "Cannot pickle a long with the size of size is larger than 1 byte"

Or, is there less-awkward-but-also-accurate preferred version of the assert text?

Comment (driazati) [reply torch/csrc/jit/pickler.cpp]: Maybe just `Cannot pickle a long larger than 255 bytes` (0 is a valid size, 256 is not)

Comment (jjlilley) [reply torch/csrc/jit/pickler.cpp]: fixed, thanks!

**Review by driazati — APPROVED** [review]

**Pull request closed without merging**